src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double DEFAULT_MULTIPLIER = 1.5;
+    private int capacity;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        capacity = DEFAULT_CAPACITY;
+        elementData = (T[]) new Object[capacity];
+    }
+
+    private void resizeIfNeeded() {
+        if (size == capacity) {
+            capacity *= DEFAULT_MULTIPLIER;
+            T[] expandedArray = (T[]) new Object[capacity];
+            System.arraycopy(elementData, 0, expandedArray, 0, size);
+            elementData = expandedArray;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of ArrayList bound!");
+        }
+    }
+
+    private boolean isEquals(Object first, Object second) {
+        return first == second || first != null && first.equals(second);
+    }
+
     @Override
     public void add(T value) {
-
+        resizeIfNeeded();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
-    }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        resizeIfNeeded();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
+    }  
 
     @Override
     public void addAll(List<T> list) {
-
+        while (size + list.size() > capacity) {
+            resizeIfNeeded();
```

Comment:
This solution will cause infinite cycle in case incoming list really requires resize as you're checking if resize needed also inside method resizeIfNeeded() and it will not know anything about how much elements will come and will still saying that size is fine

Please use your add() method in a loop and allow it determine itself if resize needed